Python manager template

elliotcmorris opened the pull request:
[#777](https://github.com/OpenAssetIO/OpenAssetIO/issues/777)

Creates a default python manager plugin project for OpenAssetIO, with linting, testing, and publishing preconfigured.

Intended to be used as a template project, and reduce the spin-up time for new managers.

The review split into 3 commits for convenience (Core, Tests, CI), but am planning to squash into a single commit as I think this should all come in together.

If this template is passable, my plan is to transfer the ownership of this repo into OpenAssetIO, which I hope is a thing you can do.

elliotcmorris commented on plugin/my_asset_manager/AssetManagerInterface.py:
```diff
@@ -0,0 +1,95 @@
+        # entity references, see :
+        # https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/blob/main/plugin/openassetio_manager_bal/BasicAssetLibraryInterface.py
+        for idx, ref in enumerate(entityReferences):
+            if ref.toString() == "my_asset_manager:///anAsset":
```
Not too sure about this. Wanted to be as direct as possible to avoid providing too many opinionated utilities in a template, but treating this as just a string makes me think twice. Would appreciate thoughts.

foundrytom replied:
I'd be tempted to leave this out, as its showing the low-level TraitsData API. We need to be steering people in the direction of MediaCreation and using well-known traits. Every custom trait is a step away from interoperability, and using the low-level API is inherently less robust. I know we use it in BAL, but that's because its deliberately a puppet. Could you replace it with a comment about this is where you bridge your native data model into OpenAssetIO traits?

If we wanted to be really helpful, documenting some of the API contract here might be good  (eg: you have to make sure you call one of the two callbacks for each index _before_ returning).

foundrytom commented on README.md:
```diff
@@ -1 +1,147 @@
-# openassetio-python-manager-template
\ No newline at end of file
+# openassetio-python-manager-template
```
[fix] We established a convention a while back for native sub-repos under OpenAssetIO where by we have increasing specificity, so when sorted, they group similar things together. ie. `OpenAssetIO-Manager-BAL` not -OpenAssetIO-BAL-Manager`, and `OpenAssetIO-Test-CMake` not `OpenAssetIO-CMake-Test`.

So I guess this would need to be something along the lines of (note the case too):
- `OpenAssetIO-Manager-Template-Python`
- `OpenAssetIO-Template-Manager-Python`
- `Template-OpenAssetIO-Manager-Python`
- :upside_down_face: 

N.B The current exceptions to this are where they follow a rule of their own parent project, eg: USD and OTIO.

elliotcmorris replied:
We'll use `Template-OpenAssetIO-Manager-Python`

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/ad40cd29fcaced4da1e0386d88c5eacf8d738a84

We'll need to remember to rename the actual repo when transferring. (or directly after merge).
There arn't any other in-source changes I don't think.

foundrytom replied:
You should be able to rename it now, GitHub should take care of redirecting old URLs

elliotcmorris replied:
Renamed

foundrytom commented on README.md:
```diff
@@ -1 +1,147 @@
\ No newline at end of file
+# openassetio-python-manager-template
+
+An [OpenAssetIO](https://github.com/OpenAssetIO/OpenAssetIO) python
```
```suggestion
An [OpenAssetIO](https://github.com/OpenAssetIO/OpenAssetIO) Python
```

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/c032106491ca07d13bda6dc65908a84f73a4d527

foundrytom commented on README.md:
```diff
@@ -1 +1,147 @@
+# openassetio-python-manager-template
+
+An [OpenAssetIO](https://github.com/OpenAssetIO/OpenAssetIO) python
+manager template.
```
```suggestion
manager plugin template.
```

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/908f091c35f429f4aee9902bcad4e94e41e36473

foundrytom commented on README.md:
```diff
@@ -1 +1,147 @@
+then an `OpenAssetIO` host can load it automatically, simply via
+configuring to the [manager identifier](plugin/my_asset_manager/__init__.py#L35),
+(usually by using a [default config file.](https://openassetio.github.io/OpenAssetIO/glossary.html#default_config_var))
+
```
[consider] Is it also worth adding a note that if you don't want to install it, you can also add the `plugin` directory to `$OPENASSETIO_PLUGIN_PATH` instead?

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/ecc0ad1a444619723b01142ec06220bdeb015f6c

foundrytom commented on plugin/my_asset_manager/AssetManagerInterface.py:
```diff
@@ -0,0 +1,95 @@
+#
```
[fix] Our usual convention would be that this would be `MyAssetManagerInterface.py` as we name single-class files with the class name.

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/031e4e50a03cdfc9ebf6aec682179fd7240b9197

foundrytom commented on plugin/my_asset_manager/AssetManagerInterface.py:
```diff
@@ -0,0 +1,95 @@
+        return "myorg.manager.my_asset_manager"
+
+    def initialize(self, managerSettings, hostSession):
+        pass
```
[fix] Sorry, this is a hangup from SAM being out of date, see: https://github.com/OpenAssetIO/OpenAssetIO/blob/41347c6b4a57d56a11724bbaa1aa1bdf661fa4ce/src/openassetio-python/package/openassetio/test/manager/apiComplianceSuite.py#L100, might need some fixtures, and this will need to raise if it's not `{}`.

elliotcmorris replied:
I think we discovered that this does the correct thing by default? Have added some doc explaining that this is the place to put your initialization logic if you have any.

foundrytom replied:
I think we worked out that its pure virtual. It would need to assert that settings are empty to be correct. I think we need to update the apiComplianceSuite to check it excepts if no valid settings are provided in the fixture.

foundrytom commented on README.md:
````diff
@@ -1 +1,147 @@
+myorg
+```
+
+with the name of your organization.
````
[fix] Clarify this should be in reverse-dns format (if its for `identifier`?)

elliotcmorris replied:
https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/b9f6fe69f6c92f2d394aa3161bc0b1411196e8d9

foundrytom commented on plugin/my_asset_manager/AssetManagerInterface.py:
```diff
@@ -0,0 +1,95 @@
+
+    def managementPolicy(self, traitSets, context, hostSession):
+        # pylint: disable=unused-argument
+        return [TraitsData() for _ in traitSets]
```
[fix] This is a critical API method. If we're going to include functioning resolve code, then we need to reflect that here, otherwise we have a manager that is saying it never manages anything, but does have data. We def don't want to encourage that. Here would be a good place to reference the MediaCreation traits, and the documentation for `managementPolicy`. (I need to update that with the last DR too!). If we don't want to go that far in the interim implementation, then we could stick with comments only? Otherwise it'd need some more test coverage/fixtures too... :rabbit: :hole:

elliotcmorris replied:
Part of the main extension work : https://github.com/elliotcmorris/openassetio-python-manager-template/commit/963920e9af7f7735c78d6aab17fb7745003f003c

foundrytom commented on plugin/my_asset_manager/__init__.py:
```diff
@@ -0,0 +1,60 @@
+#
+
+"""
+Module Documentation for MyAssetManager here.
```
[consider] You could probably be more specific here, ie something along the lines of how this exports the interface implementation as a valid OpenAssetIO plugin.

elliotcmorris replied:
I lean towards leaving this, as I think i'm correct in saying that the module documentation for the init file is the place for a summary of the module as a whole. I'd expect the user of this template to replace any string we put here in totality, to make it appropriate for their end users. I think the description of this file in `README` is the better place for that sort of explainer.

foundrytom commented on tests/fixtures.py:
```diff
@@ -0,0 +1,33 @@
+
+KIDENTIFIER = "myorg.manager.my_asset_manager"
+
+fixtures = {
```
[fix] If were going to include resolve in the implementation, we should also include fixtures here too for https://github.com/OpenAssetIO/OpenAssetIO/blob/41347c6b4a57d56a11724bbaa1aa1bdf661fa4ce/src/openassetio-python/package/openassetio/test/manager/apiComplianceSuite.py#L278, same for `managementPolicy` if we do something there.

elliotcmorris replied:
Get rid of the K if upper cased.

elliotcmorris replied:
I added fixtures for `Test_resolve`

The ManagementPolicy compliance tests don't require any fixtures as far as I can tell. They're running without them (I can tell because changing the management policy implementation to be incorrect causes a failure)

elliotcmorris replied:
Part of the main extension work : https://github.com/elliotcmorris/openassetio-python-manager-template/commit/963920e9af7f7735c78d6aab17fb7745003f003c

foundrytom reviewed (changes requested):
Great stuff @elliotcmorris this is going to be _so_ helpful for those new to the space! Thanks for taking the initiative on this!

I've asked legal about what we should have for the copyright, as a template, get back to you on that soon.

I guess the main Q is deciding how far you want to go with sample implementation, as if we do have working stuff like `resolve` it does need a little more to go with it (`managementPolicy`, other fixtures). If we want to keep the scope smaller, we can leave it non-functional with helpful comments, but then that does undo some of the loveliness you did with the test coverage.

foundrytom commented:
@elliotcmorris Whilst we're working out the right way to license this, would you be able to remote license comments from any of the files other than `LICENSE` please?

elliotcmorris commented:
@foundrytom 

As we discussed, have evolved this template to be more of a legitimate and full demonstration, along with clarifying inline documentation : https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/963920e9af7f7735c78d6aab17fb7745003f003c

Also removed license comments: https://github.com/elliotcmorris/openassetio-python-manager-template/pull/2/commits/badea329cd7cfb083ff5ba53644050055571d5d5

History is getting a little messy, but just to be clear, I'm planning to squash this all down into one commit prior to merge.

feltech reviewed (approved):
Looks good to me, will be a really useful reference.

elliotcmorris committed:
[Core] Initial Commit

Create a template repository for python manager plugins.
This should drastically cut down the work required to startup a new
python manager, and avoid re-arbitrating lint/CI/deployment
paradigms.

Signed-off-by: Elliot Morris <elliot.morris@foundry.com>

elliotcmorris commented:
Did an over-the-shoulder roundup review with Tom, despite him not having hit the button yet, am merging.
